components/style/stylesheet_set.rs

```diff
@@ -0,0 +1,105 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! A centralized set of stylesheets for a document.
+
+use arc_ptr_eq;
+use std::sync::Arc;
+use stylesheets::Stylesheet;
+
+/// The set of stylesheets effective for a given document.
+pub struct StylesheetSet {
+    /// The actual list of all the stylesheets that apply to the given document.
+    ///
+    /// This is only a list of top-level stylesheets, and as such it doesn't
+    /// include recursive `@import` rules.
+    stylesheets: Vec<Arc<Stylesheet>>,
+
+    /// Whether the stylesheets list above has changed since the last restyle.
+    dirty: bool,
+
+    /// Has author style been disabled?
+    author_style_disabled: bool,
+}
+
+impl StylesheetSet {
+    /// Create a new empty StylesheetSet.
+    pub fn new() -> Self {
+        StylesheetSet {
+            stylesheets: vec![],
+            dirty: false,
+            author_style_disabled: false,
+        }
+    }
+
+    /// Returns whether author styles have been disabled for the current
+    /// stylesheet set.
+    pub fn author_style_disabled(&self) -> bool {
+        self.author_style_disabled
+    }
+
+    fn remove_stylesheet_if_present(&mut self, sheet: &Arc<Stylesheet>) {
+        self.stylesheets.retain(|x| !arc_ptr_eq(x, sheet));
+    }
+
+    /// Append a new stylesheet to the current set.
+    pub fn append_stylesheet(&mut self, sheet: &Arc<Stylesheet>) {
+        self.remove_stylesheet_if_present(sheet);
+        self.stylesheets.push(sheet.clone());
+        self.dirty = true;
+    }
+
+    /// Prepend a new stylesheet to the current set.
+    pub fn prepend_stylesheet(&mut self, sheet: &Arc<Stylesheet>) {
+        self.remove_stylesheet_if_present(sheet);
+        self.stylesheets.insert(0, sheet.clone());
+        self.dirty = true;
+    }
+
+    /// Insert a given stylesheet before another stylesheet in the document.
+    pub fn insert_stylesheet_before(&mut self,
+                                    sheet: &Arc<Stylesheet>,
+                                    before: &Arc<Stylesheet>) {
+        self.remove_stylesheet_if_present(sheet);
+        let index = self.stylesheets.iter().position(|x| {
+            arc_ptr_eq(x, before)
+        }).expect("`before` stylesheet not found");
+        self.stylesheets.insert(index, sheet.clone());
+        self.dirty = true;
+    }
+
+    /// Remove a given stylesheet from the set.
+    pub fn remove_stylesheet(&mut self, sheet: &Arc<Stylesheet>) {
+        self.remove_stylesheet_if_present(sheet);
+        self.dirty = true;
```

Comment:
Would it be a win to only set dirty if we actually removed something? Or are we basically always going to remove something?